safehttp/mux_test.go

```diff
@@ -0,0 +1,204 @@
+			wantStatus: 404,
+			wantHeader: map[string][]string{
+				"Content-Type":           {"text/plain; charset=utf-8"},
+				"X-Content-Type-Options": {"nosniff"},
+			},
+			wantBody: "Not Found\n",
+		},
+		{
+			name:       "Invalid Method",
+			req:        httptest.NewRequest("POST", "http://foo.com/", nil),
+			wantStatus: 405,
+			wantHeader: map[string][]string{
+				"Content-Type":           {"text/plain; charset=utf-8"},
+				"X-Content-Type-Options": {"nosniff"},
+			},
+			wantBody: "Method Not Allowed\n",
+		},
+	}
+
+	for _, tt := range test {
+		t.Run(tt.name, func(t *testing.T) {
+			mux := NewServeMux(testDispatcher{}, "foo.com")
+
+			h := HandlerFunc(func(w ResponseWriter, r *IncomingRequest) Result {
+				return w.Write(safehtml.HTMLEscaped("<h1>Hello World!</h1>"))
+			})
+			mux.Handle("/", MethodGet, h)
+
+			b := &strings.Builder{}
+			rw := newResponseRecorder(b)
+
+			mux.ServeHTTP(rw, tt.req)
+
+			if rw.status != tt.wantStatus {
+				t.Errorf("rw.status: got %v want %v", rw.status, tt.wantStatus)
+			}
+
+			if diff := cmp.Diff(tt.wantHeader, map[string][]string(rw.header)); diff != "" {
+				t.Errorf("rw.header mismatch (-want +got):\n%s", diff)
+			}
+
+			if got := b.String(); got != tt.wantBody {
+				t.Errorf("response body: got %q want %q", got, tt.wantBody)
+			}
+		})
+	}
+}
+
+func TestHandleTwoMethods(t *testing.T) {
+	d := testDispatcher{}
+	mux := NewServeMux(d, "foo.com")
+
+	registeredGetHandler := HandlerFunc(func(w ResponseWriter, r *IncomingRequest) Result {
+		return w.Write(safehtml.HTMLEscaped("<h1>Hello World! GET</h1>"))
+	})
+	registeredPostHandler := HandlerFunc(func(w ResponseWriter, r *IncomingRequest) Result {
+		return w.Write(safehtml.HTMLEscaped("<h1>Hello World! POST</h1>"))
+	})
+	mux.Handle("/bar", MethodGet, registeredGetHandler)
+	mux.Handle("/bar", MethodPost, registeredPostHandler)
+
+	postReq := httptest.NewRequest("POST", "http://foo.com/bar", nil)
+	b := &strings.Builder{}
+	rw := newResponseRecorder(b)
+
+	mux.ServeHTTP(rw, postReq)
+
+	if want := 200; rw.status != want {
+		t.Errorf("rw.status: got %v want %v", rw.status, want)
+	}
+
+	if diff := cmp.Diff(map[string][]string{}, map[string][]string(rw.header)); diff != "" {
+		t.Errorf("rw.header mismatch (-want +got):\n%s", diff)
+	}
+
+	if got, want := b.String(), "&lt;h1&gt;Hello World! POST&lt;/h1&gt;"; got != want {
+		t.Errorf("response body: got %q want %q", got, want)
+	}
+
+	getReq := httptest.NewRequest("GET", "http://foo.com/bar", nil)
+	b = &strings.Builder{}
+	rw = newResponseRecorder(b)
+
+	mux.ServeHTTP(rw, getReq)
+
+	if want := 200; rw.status != want {
+		t.Errorf("rw.status: got %v want %v", rw.status, want)
+	}
+
+	if diff := cmp.Diff(map[string][]string{}, map[string][]string(rw.header)); diff != "" {
+		t.Errorf("rw.header mismatch (-want +got):\n%s", diff)
+	}
+
+	if got, want := b.String(), "&lt;h1&gt;Hello World! GET&lt;/h1&gt;"; got != want {
+		t.Errorf("response body: got %q want %q", got, want)
+	}
```

Comment:
Why not a tabletest?

Reply:
Refactored to tabletest in [3b8ef82](url)